File: src/main/java/page/info/EnemyInfoPage.java
package page.info;

import common.system.ENode;
import page.JBTN;
import page.JTG;
import page.Page;
import page.view.EnemyViewPage;

import javax.swing.*;

public class EnemyInfoPage extends Page {

	private static final long serialVersionUID = 1L;

	private final JBTN back = new JBTN(0, "back");
	private final JBTN anim = new JBTN(0, "anim");
	private final JBTN prev = new JBTN(0, "prev");
	private final JBTN next = new JBTN(0, "next");
	private final JBTN find = new JBTN(0, "stage");
	private final JTG extr = new JTG(0, "extra");
	private final JLabel source = new JLabel("Source of enemy icon: DB");
	private final JPanel cont = new JPanel();
	private final JScrollPane jsp = new JScrollPane(cont);
	private final EnemyInfoTable info;
	private final TreaTable trea;

	private final ENode e;

	public EnemyInfoPage(Page p, ENode de) {
		super(p);
		e = de;

		info = new EnemyInfoTable(this, de.val, de.mul, de.mula);
		trea = new TreaTable(this);

		ini();
	}

	@Override
	protected JButton getBackButton() {
		return back;
	}

	@Override
	public void callBack(Object o) {
		info.reset();
	}

	@Override
	protected void resized(int x, int y) {
		setBounds(0, 0, x, y);

		set(back, x, y, 0, 0, 200, 50);
		set(source, x, y, 0, 50, 600, 50);
		set(prev, x, y, 300, 0, 200, 50);
		set(anim, x, y, 600, 0, 200, 50);
		set(next, x, y, 900, 0, 200, 50);
		set(find, x, y, 1200, 0, 200, 50);
		set(extr, x, y, 1500, 0, 200, 50);
		set(jsp, x, y, 50, 100, 1650, 1150);
		set(trea, x, y, 1700, 100, 400, 1200);

		int ih = info.getH();

		cont.setPreferredSize(size(x, y, 1600, ih - 50).toDimension());

		jsp.getVerticalScrollBar().setUnitIncrement(size(x, y, 50));

		set(info, x, y, 0, 0, 1600, ih);
	}

	@Override
	public synchronized void onTimer(int t) {
		super.onTimer(t);

		jsp.revalidate();
	}

	private void addListeners() {
		back.addActionListener(arg0 -> changePanel(getFront()));

		prev.addActionListener(arg0 -> changePanel(new EnemyInfoPage(getFront(), (ENode) e.prev)));

		anim.addActionListener(arg0 -> {
			if (getFront() instanceof EnemyViewPage)
				changePanel(getFront());
			else
				changePanel(new EnemyViewPage(getThis(), e.val));
		});

		next.addActionListener(arg0 -> changePanel(new EnemyInfoPage(getFront(), (ENode) e.next)));

		find.addActionListener(arg0 -> changePanel(new StageFilterPage(getThis(), e.val.findApp())));

		extr.addActionListener(arg0 -> {
			info.setDisplaySpecial(extr.isSelected());

			fireDimensionChanged();
		});
	}

	private void ini() {
		assignSubPage(info);

		add(back);

		cont.add(info);
		cont.setLayout(null);

		add(jsp);
		add(trea);
		add(prev);
		add(anim);
		add(next);
		add(find);
		add(extr);
		add(source);

		prev.setEnabled(e.prev != null);
		next.setEnabled(e.next != null);

		addListeners();
	}

}

File: src/page/info/EnemyInfoPage.kt
package page.info

import common.battle.BasisSet
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.Enemy
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JBTN
import page.JL
import page.Page
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.EnemyViewPage
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.JLabel

class EnemyInfoPage @JvmOverloads constructor(p: Page?, de: Enemy, mul: Int = 100, mula: Int = 100) : Page(p) {
    private val back: JBTN = JBTN(0, "back")
    private val anim: JBTN = JBTN(0, "anim")
    private val find: JBTN = JBTN(0, "stage")
    private val source = JLabel("Source of enemy icon: DB")
    private val info: EnemyInfoTable
    private val trea: TreaTable
    private val e: Enemy
    private val b: BasisSet = BasisSet.Companion.current()
    override fun callBack(o: Any?) {
        info.reset()
    }

    override fun resized(x: Int, y: Int) {
        setBounds(0, 0, x, y)
        Page.Companion.set(back, x, y, 0, 0, 200, 50)
        Page.Companion.set(source, x, y, 0, 50, 600, 50)
        Page.Companion.set(anim, x, y, 600, 0, 200, 50)
        Page.Companion.set(find, x, y, 1200, 0, 200, 50)
        Page.Companion.set(info, x, y, 50, 100, 1600, 800)
        Page.Companion.set(trea, x, y, 1700, 100, 400, 1200)
    }

    private fun addListeners() {
        back.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                if (front is StageFilterPage) changePanel(front.getFront()) else changePanel(front)
            }
        })
        anim.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                if (front is EnemyViewPage) changePanel(front) else changePanel(EnemyViewPage(getThis(), e))
            }
        })
        find.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                changePanel(StageFilterPage(getThis(), e.findApp()))
            }
        })
    }

    private fun ini() {
        add(back)
        add(info)
        add(trea)
        add(anim)
        add(find)
        add(source)
        addListeners()
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        e = de
        info = EnemyInfoTable(this, de, mul, mula)
        trea = TreaTable(this, b)
        ini()
        resized()
    }
}
